### DISCUSSION TITLE
Why do rebuilds take so long whenever flags change?

### DISCUSSION THREAD

**Discussion by georgefst** [haskell/cabal#9302] [General]
I've ended up here after investigating some practical issues with using HLS and Cabal, prompted by [a Discourse thread](https://discourse.haskell.org/t/convenience-in-the-haskell-ecosystem/7731/77) about papercuts in the ecosystem. It's become clear to me that the amount of time Cabal takes to rebuild a project after fairly trivial changes to its `.cabal` file is a bottleneck.

Adding a module under `other_modules`, or a dependency compatible with the existing set (no solving!) causes rebuilds to take about 10s on my fairly-powerful machine. This is particularly annoying in the early stages of a project where dependencies may be frequently added and removed. For comparison, a `cabal build` after a trivial change to a source file, without any config changes, can take as little as 0.5s. And a complete no-op build (`cabal build && time cabal build`) takes around 0.02s.

Before I get in to profiling this myself, I'm just wondering whether any maintainers have a succinct explanation of what's taking all that time, or know of any quick wins?

**Comment by georgefst** [comment]
Huh, I just thought to run `cabal build -v`, and to my surprise it looks like most of the time is spent calling `pkg-config`. Will investigate further...

**Reply by geekosaur** [reply]
See also https://github.com/haskell/cabal/issues/8930

**Reply by gbaz** [reply]
Note that the solver is _always_ run when a cabal file is changed, because we cache the _build plan_ and not individual components of it. Its only in running the solver that the pkg-config call(s) will be made. It would be a small change (if tricky to find exactly where to make it) to cache and key the solver results specifically on the requested packages, so that they could stay stable under "safe" changes to the cabal file.

**Reply by georgefst** [reply]
As for `pkg-config` slowness, I've posted a suggestion in https://github.com/haskell/cabal/issues/8930#issuecomment-1744770304.

> It would be a small change (if tricky to find exactly where to make it) to cache and key the solver results specifically on the requested packages, so that they could stay stable under "safe" changes to the cabal file.

This is an interesting proposition regardless.

**Comment by ulysses4ever** [comment]
I think we can close this and move the suggestions in the above mentioned tickets. Thanks everyone!
